// resume-pdf/lib/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var jspdf_1 = require("jspdf");
var utils_1 = require("./utils");
var ResumePDF = function (content) {
    var config = SetDocConfig(content);
    var _a = config.content, headerConfig = _a.header, summaryConfig = _a.summary, topSkillsConfig = _a.top_skills, experienceConfig = _a.experience, educationConfig = _a.education;
    var resume = new jspdf_1.default({ unit: 'in' });
    resume.setLineWidth(0.005);
    utils_1.AddHeader(resume, config, headerConfig);
    utils_1.AddSummary(resume, config, summaryConfig);
    utils_1.AddTopSkills(resume, config, topSkillsConfig);
    utils_1.AddExperience(resume, config, experienceConfig);
    utils_1.AddEducation(resume, config, educationConfig);
    return resume;
};
function SetDocConfig(content) {
    var result = {
        margins: {
            top: 1,
            left: 1,
            right: 1,
            bottom: 1,
        },
        pageSize: {
            width: 8.25,
            height: 11.75,
        },
        fontConfig: {
            titleSize: 16,
            textSize: 11,
        },
        sectionTitleSpacing: 0.4,
        sectionTextSpacing: 0.3,
        sectionEndSpacing: 0.2,
        currentY: 0,
        content: content,
    };
    return result;
}
exports.default = ResumePDF;
